### PR TITLE
Use strict typescript option in audits module in Angular

### PR DESCRIPTION
**Pull request by kaidohallik**
This is next peace for #10631

Some comments about changes.

* Switching on typescript `strict` option forces to look at all kind of corner cases. So I made some changes in logic. I tested all these cases manually, but added also full unit tests.

* For date testing added `jest-date-mock` library.

* All internal methods are now private.

* Error messages were not shown before these changes.

* If start or end date was empty then NPE in server side before these changes.

* I removed `routeData.unsubscribe()`, because this in not necessary according to https://angular.io/guide/router#!#route-parameters (but it states: _Feel free to unsubscribe anyway. It is harmless and never a bad practice._) - currently in `audits` and `user-management` this `unsubscribe` is used, but in `entities` is not used. I believe it's better to not write unnecessary code as this is currently in entities.

* `link` header is not used, so removed variable for that

-   Please make sure the below checklist is followed for Pull Requests.

-   [x] [Travis tests](https://travis-ci.org/jhipster/generator-jhipster/pull_requests) are green
-   [x] Tests are added where necessary
-   [x] Documentation is added/updated where necessary
-   [x] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/master/CONTRIBUTING.md) are followed

### CODE REVIEW

File: generators/client/templates/angular/src/main/webapp/app/admin/audits/audits.component.ts.ejs [hunk 1]
```diff
@@ -30,113 +29,109 @@ import { AuditsService } from './audits.service';
...
-    links: any;
-    page: number;
-    routeData: any;
-    predicate: any;
-    previousPage: any;
-    reverse: boolean;
-    toDate: string;
-    totalItems: number;
+export class AuditsComponent implements OnInit {
+    private dateFormat = 'yyyy-MM-dd';
```
Comment (wmarques) [inline generators/client/templates/angular/src/main/webapp/app/admin/audits/audits.component.ts.ejs]: I would put private fields at the end of public fields

File: generators/client/templates/angular/src/main/webapp/app/admin/audits/audits.component.ts.ejs [hunk 2]
```diff
@@ -30,113 +29,109 @@ import { AuditsService } from './audits.service';
...
         }
-
-        this.fromDate = this.datePipe.transform(fromDate, dateFormat);
+        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+        this.fromDate = this.datePipe.transform(date, this.dateFormat)!;
     }
 
-    today() {
-        const dateFormat = 'yyyy-MM-dd';
+    private today(): void {
```
Comment (wmarques) [inline generators/client/templates/angular/src/main/webapp/app/admin/audits/audits.component.ts.ejs]: I would prefer this function returning today + 1 Date and then you can do this.toDate = this.today()
WDYT ?

File: generators/client/templates/angular/src/main/webapp/app/admin/audits/audits.component.ts.ejs [hunk 3]
```diff
@@ -30,113 +29,109 @@ import { AuditsService } from './audits.service';
...
-        this.alertService.error(error.error, error.message, null);
+    private onSuccess(audits: Audit[] | null, headers: HttpHeaders): void {
+        this.totalItems = Number(headers.get('X-Total-Count'));
+        if (audits === null) {
+            this.audits = [];
+        } else {
+            this.audits = audits;
+        }
+        this.dataHaveBeenLoaded = true;
+        this.initialized = true;
```
Comment (wmarques) [inline generators/client/templates/angular/src/main/webapp/app/admin/audits/audits.component.ts.ejs]: What's the difference between dataHaveBeenLoaded and initialized ?
Btw it's dataHasBeenLoaded I think 😄

File: generators/client/templates/angular/src/main/webapp/app/admin/audits/audits.service.ts.ejs [hunk 4]
```diff
@@ -28,10 +28,8 @@ import { Audit } from './audit.model';
 export class AuditsService  {
     constructor(private http: HttpClient) { }
 
-    query(req: any): Observable<HttpResponse<Audit[]>> {
+    query(req: { page?: number; size?: number; sort?: string[]; fromDate: string; toDate: string }): Observable<HttpResponse<Audit[]>> {
         const params: HttpParams = createRequestOption(req);
-        params.set('fromDate', req.fromDate);
-        params.set('toDate', req.toDate);
```
Comment (wmarques) [inline generators/client/templates/angular/src/main/webapp/app/admin/audits/audits.service.ts.ejs]: Why are we removing thos params from the request?

File: generators/client/templates/angular/src/main/webapp/app/admin/audits/audits.component.ts.ejs (hunk 1 above)
Comment (kaidohallik) [reply generators/client/templates/angular/src/main/webapp/app/admin/audits/audits.component.ts.ejs]: Done.

File: generators/client/templates/angular/src/main/webapp/app/admin/audits/audits.component.ts.ejs (hunk 2 above)
Comment (kaidohallik) [reply generators/client/templates/angular/src/main/webapp/app/admin/audits/audits.component.ts.ejs]: Good idea, done and made the same also for the `previousMonth()` function.

File: generators/client/templates/angular/src/main/webapp/app/admin/audits/audits.component.ts.ejs (hunk 3 above)
Comment (kaidohallik) [reply generators/client/templates/angular/src/main/webapp/app/admin/audits/audits.component.ts.ejs]: `initialized` is valued to `true` also then if error occurs on data loading. So if on first page load error occurs then anyway date filter is shown and user can try to change dates and retry data loading in this way.

File: generators/client/templates/angular/src/main/webapp/app/admin/audits/audits.service.ts.ejs (hunk 4 above)
Comment (kaidohallik) [reply generators/client/templates/angular/src/main/webapp/app/admin/audits/audits.service.ts.ejs]: Those parameters are not removed from the request. These parameters are already assigned in the first line of this function and no need to assign values again.

File: generators/client/templates/angular/src/main/webapp/app/admin/audits/audits.component.ts.ejs (hunk 3 above)
Comment (kaidohallik) [reply generators/client/templates/angular/src/main/webapp/app/admin/audits/audits.component.ts.ejs]: I moved `initialized = true` to `ngOnInit` now.  
It aligns initial behavior with further behavior: running new query before previous query is not completed is now allowed also while first query is running.  
And code looks better after this change.  
I force pushed this change.

File: generators/client/templates/angular/src/main/webapp/app/admin/audits/audits.component.ts.ejs [hunk 5]
```diff
@@ -30,113 +29,110 @@ import { AuditsService } from './audits.service';
...
+    itemsPerPage = ITEMS_PER_PAGE;
+    page!: number;
+    predicate!: string;
+    previousPage!: number;
+    ascending!: boolean;
+    toDate = '';
+    totalItems = 0;
+    canLoad = false;
+    initialized = false;
+    dataHaveBeenLoaded = false;
```
Comment (wmarques) [inline generators/client/templates/angular/src/main/webapp/app/admin/audits/audits.component.ts.ejs]: I'm not a big fan of those 3 booleans... Can we instead use a `state` variable with 3 values `'initialized' | 'loading' | 'loaded'` or something like this ? WDYT ?

File: generators/client/templates/angular/src/main/webapp/app/admin/audits/audits.component.html.ejs [hunk 6]
```diff
@@ -36,17 +38,24 @@
         </div>
     </div>
 
-    <div class="alert alert-warning" *ngIf="audits?.length === 0">
+    <div class="alert alert-warning" *ngIf="dataHaveBeenLoaded && audits.length === 0">
```
Comment (wmarques) [inline generators/client/templates/angular/src/main/webapp/app/admin/audits/audits.component.html.ejs]: Typo: `dataHasBeenLoaded`

File: generators/client/templates/angular/src/main/webapp/app/admin/audits/audits.component.ts.ejs (hunk 5 above)
Comment (kaidohallik) [reply generators/client/templates/angular/src/main/webapp/app/admin/audits/audits.component.ts.ejs]: Thanks for pointing this out! I removed all 3 variables:
* canLoad --> now it's function, value is calculated
* initialized --> just removed, filter section survives if this is shown before `ngOnInit`
* dataHaveBeenLoaded --> `audits` array variable can now be also `null`, null means that data not loaded yet

I also removed `eslint disable` lines and set `@typescript-eslint/no-non-null-assertion` off as decided in #10639 - hopefully Git conflicts will not arise if merging different PR-s (exactly the same change in `.eslintrc`).

I amended previous review commit and force pushed.

File: generators/client/templates/angular/src/main/webapp/app/admin/audits/audits.component.html.ejs (hunk 6 above)
Comment (kaidohallik) [reply generators/client/templates/angular/src/main/webapp/app/admin/audits/audits.component.html.ejs]: variable now removed at all

**Comment by kaidohallik** [comment]
As per https://github.com/jhipster/generator-jhipster/pull/10639#discussion_r348303798 I used `?:` in `audits` array declaration. I amended latest commit and force pushed.

**Comment by wmarques** [comment]
Let's go ahead and see how it goes ! Again, thanks @kaidohallik 👏 
That's a huge work and it provides a much better Angular code 😄

**Comment by kaidohallik** [comment]
@wmarques thank you for your reviews!  
These helped to make these PR-s better.

**Comment by wmarques** [comment]
Well it took time but I managed to do it :)
